Feature/jra 26 android register user

Rodrigo-Schar opened the pull request:
- Added Register User
- Added Firebase Auth
- Added Email validation
- added hook and fix request changes
- added date to Double, observers with viewModel
- merge and fix model
- Added UnitTest for validateFields

jmacboy commented on app/src/main/java/com/jalasoft/routesapp/data/remote/managers/FirebaseManager.kt:
```diff
@@ -0,0 +1,40 @@
+        fun getUsersByParameter(collection: FirebaseCollections, field: String, parameter: String, successListener: (MutableList<User>) -> Unit, errorListener: (String) -> Unit) {
+
+            db.collection(collection.toString()).whereEqualTo(field, parameter).get().addOnSuccessListener { documents ->
+                //var users = mutableListOf<User>()
```
Remove unnecessary comments

jmacboy commented on app/src/main/java/com/jalasoft/routesapp/data/remote/managers/UserManager.kt:
```diff
@@ -0,0 +1,39 @@
+import com.jalasoft.routesapp.util.helpers.UserTypeLogin
+
+class UserManager {
+    object Singleton {
```
There’s no need to declare a Singleton object, if you’re just going to use static methods, just change class UserManager for object UserManager

jmacboy commented on app/src/main/java/com/jalasoft/routesapp/ui/auth/login/view/LoginFragment.kt:
```diff
@@ -0,0 +1,32 @@
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        // Inflate the layout for this fragment
```
Remove unnecessary comments

jmacboy commented on app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/view/RegisterUserFragment.kt:
```diff
@@ -0,0 +1,122 @@
+            val password = binding.etRegPassword.text.toString()
+
+            viewModel.registerUserAuth(name, email, password, { _ ->
+                binding.txtRegError.isVisible = false
```
This is wrong, success and failure should be handled on the viewmodel and update variables, here you just need to listen to variable changes

Rodrigo-Schar replied:
Done!

jmacboy replied:
This is still wrong, this logic should be handled on the viewmodel, and you should be observing variable changes on the Fragment. I don't see a place where this is happening, you´re trying to imitate that with the callback but it isn't that way.

![imagen](https://user-images.githubusercontent.com/1596425/183504207-cb16ce32-f4ee-4fca-91b0-7f61afc3199b.png)

Example from this link:
https://developer.android.com/topic/libraries/architecture/livedata#observe_livedata_objects

Rodrigo-Schar replied:
i think now it is good i add the observer for the viewModel you can check it

jmacboy commented on app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/view/RegisterUserFragment.kt:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    private fun validateFields(): Boolean {
+        var isValid = true
```
Method is good but the logic should be moved to the viewmodel and update the variables in there. When the variables are updated refresh the fragment here

Rodrigo-Schar replied:
Done!

jmacboy commented on app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/viewModel/RegisterUserViewModel.kt:
```diff
@@ -0,0 +1,48 @@
+        if (validateEmail(email)) {
+            UserManager.Singleton.createUserAuth(email, password, { _ ->
+                registerUser(name, email, { success ->
+                    successListener(success)
```
The logic should be handled here

jmacboy commented on app/src/main/res/layout/fragment_login.xml:
```diff
@@ -0,0 +1,22 @@
+    android:layout_height="match_parent"
+    tools:context=".ui.auth.login.view.LoginFragment">
+
+    <!-- TODO: Update blank fragment layout -->
```
Remove unnecessary comments

jmacboy commented on app/src/main/res/values/themes.xml:
```diff
@@ -13,4 +13,18 @@
+    <style name="btnPrimaryRounded" parent="Widget.MaterialComponents.Button.OutlinedButton">
+        <item name="shapeAppearanceOverlay">@style/MyShapeAppearance</item>
+    </style>
+    <style name="MyShapeAppearance">
```
Add a more descriptive name to this appearance

YawarValeriano reviewed (approved):
Seems good to me

YawarValeriano commented on app/src/main/java/com/jalasoft/routesapp/data/model/remote/User.kt:
```diff
@@ -0,0 +1,12 @@
+package com.jalasoft.routesapp.data.model.remote
+
+data class User(
+    val Id: String? = null,
```
We should define naming variables form (camelCase, PascalCase)

YawarValeriano commented:
Also pull changes from develop to add de sonar gradle config for this pull request

jmacboy commented on app/src/main/java/com/jalasoft/routesapp/data/model/remote/User.kt:
```diff
@@ -0,0 +1,12 @@
+    val type: Int? = null,
+    val typeLogin: Int? = null,
+    val createdAt: Double? = null,
+    val updatedAt: Double? = null
```
Does the date has decimals? or why double instead of long

Rodrigo-Schar replied:
because in ios it is saving with decimals so to don't get issues

jmacboy reviewed (changes requested):
the other stuff is ok now, just one question this time

jmacboy commented:
Also I see 0 unit tests on this code @Rodrigo-Schar

jmacboy commented:
just missing unit tests and you´re good to go

Rodrigo-Schar commented:
i changed my validateFuntion to try to don't test the liveData, i tried with others options to test the liveData but nothing works , tomorrow Alvaro will show an example of how to test with mock so i'm not sure if i can pull this request with that two unitest that i added

jmacboy reviewed (approved):
I´m approving it by now, so you can merge, but you really need to fix the unit test issues
